### DISCUSSION TITLE
Why does .setData() ask for a void query key?

### DISCUSSION THREAD

**Discussion by yogaraptor** [trpc/trpc#4376] [Q&A] [answered]
My very simple use-case:

- Fire `addToCart` mutation
- On success, add the new cart response to the cache for the `getCart` query

This code works great:

```ts
  const utils = trpc.useContext();
  const addToCart = trpc.cart.add.useMutation({
    onSuccess(data) {
      utils.cart.getMini.setData(undefined, data);
    },
  });
```

but having to add the `undefined` for the query key bugs me. I realise that there's no need to provide it, since tRPC can generate it from the specific mutation/query being called, but why does tRPC ask for it in the first place? Are there situations where you'd need to provide it? Screenshot of the type below:

![A screenshot of a type definition for setData in the above code snippet. It shows the type of the first parameter to setData (input) as "void"](https://github.com/trpc/trpc/assets/1737376/56950f3e-9d10-4410-812c-336217c146cf)

**Comment by c-ehrlich** [comment, marked as answer]
The use case is when your queryKey includes an input, such as:

```ts
utils.cart.getMini.setData({ id: 1 }, data);
```

I do agree that a more elegant api might be:
```ts
utils.cart.getMini.setData({
  input: { id: 1 }, // no input added to queryKey if this is not passed
  data,
});
```

React Query is heading towards a similar object syntax in many parts of their API in v5, so this might already be planned anyway?

**Reply by yogaraptor** [reply]
Makes sense, thanks so much!
